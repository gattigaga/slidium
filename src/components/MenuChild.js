import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Icon from "components/Icon";

const Container = styled.div`
  width: 100%;
  padding: 8px 16px;
  box-sizing: border-box;
  cursor: ${({ isClickable }) => (isClickable ? "pointer" : "default")};
  position: relative;
  display: flex;
  align-items: center;
  user-select: none;

  &:hover {
    ${({ isClickable }) => isClickable && "background: #eee;"};
  }
`;

const IconContainer = styled.div`
  width: 32px;
  display: flex;
  position: relative;
`;

const StyledIcon = styled(Icon)`
  font-size: 20px;
  color: ${({ isClickable }) => (isClickable ? "#888" : "#bbb")};
  position: absolute;
  top: 50%;
  left: 0%;
  transform: translateY(-50%);
`;

const Label = styled.span`
  font-size: 12px;
  font-family: Roboto;
  color: ${({ isClickable }) => (isClickable ? "black" : "#bbb")};
`;

const MenuChild = ({ label, icon, isClickable, onClick }) => (
  <Container isClickable={isClickable} onClick={() => isClickable && onClick()}>
    <IconContainer>
      {icon && <StyledIcon name={icon} isClickable={isClickable} />}
    </IconContainer>
    <Label isClickable={isClickable}>{label}</Label>
  </Container>
);

MenuChild.propTypes = {
  icon: PropTypes.string,
  label: PropTypes.string,
  isClickable: PropTypes.bool,
  onClick: PropTypes.func
};

MenuChild.defaultProps = {
  label: "Menu",
  isClickable: true,
  onClick: () => {}
};

export default MenuChild;
