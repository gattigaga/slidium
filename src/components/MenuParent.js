import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  width: auto;
  padding: 4px 8px;
  cursor: default;
  position: relative;
  user-select: none;

  &:hover {
    background: #eee;
  }
`;

const Label = styled.span`
  font-size: 12px;
  font-family: Roboto;
  color: black;
`;

const LabelActive = Container.extend`
  position: absolute;
  top: -28px;
  left: -1px;
  z-index: 3;
  background: white;
  border: 1px solid #ccc;
  border-bottom: 0px;
  padding-bottom: 5px;

  &:hover {
    background: white;
  }
`;

const Tray = styled.span`
  position: absolute;
  min-width: 192px;
  top: 100%;
  left: 0%;
  background: white;
  border: 1px solid #ccc;
  padding: 8px 0px;
  box-shadow: 1px 1px 8px rgba(0, 0, 0, 0.2);
  z-index: 2;
`;

const MenuParent = ({ label, children, isOpen, onClick }) => (
  <Container onClick={onClick}>
    <Label>{label}</Label>
    {isOpen && (
      <Tray>
        <LabelActive>
          <Label>{label}</Label>
        </LabelActive>
        {children}
      </Tray>
    )}
  </Container>
);

MenuParent.propTypes = {
  label: PropTypes.string,
  children: PropTypes.node,
  onClick: PropTypes.func,
  isOpen: PropTypes.bool
};

MenuParent.defaultProps = {
  label: "Menu"
};

export default MenuParent;
