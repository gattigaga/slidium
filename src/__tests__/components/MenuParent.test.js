import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import MenuParent from "components/MenuParent";
import MenuChild from "components/MenuChild";

let wrapper, onClick;

beforeEach(() => {
  onClick = jest.fn();
  wrapper = shallow(<MenuParent label="File" onClick={onClick} />);
});

describe("MenuParent", () => {
  it("renders in close mode", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders in open mode without children", () => {
    wrapper.setProps({ isOpen: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders in open mode with children", () => {
    wrapper = shallow(
      <MenuParent label="File" onClick={onClick}>
        <MenuChild label="New" />
        <MenuChild label="Open" isClickable={false} />
        <MenuChild label="Print" icon="print" />
      </MenuParent>
    );

    wrapper.setProps({ isOpen: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("calls 'onClick'", () => {
    expect(onClick.mock.calls.length).toEqual(0);
    wrapper.simulate("click");
    expect(onClick.mock.calls.length).toEqual(1);
  });
});
