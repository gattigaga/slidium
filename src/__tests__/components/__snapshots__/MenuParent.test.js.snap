// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuParent renders in close mode 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.span>
    File
  </styled.span>
</styled.div>
`;

exports[`MenuParent renders in open mode with children 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.span>
    File
  </styled.span>
  <styled.span>
    <styled.div>
      <styled.span>
        File
      </styled.span>
    </styled.div>
    <MenuChild
      isClickable={true}
      label="New"
      onClick={[Function]}
    />
    <MenuChild
      isClickable={false}
      label="Open"
      onClick={[Function]}
    />
    <MenuChild
      icon="print"
      isClickable={true}
      label="Print"
      onClick={[Function]}
    />
  </styled.span>
</styled.div>
`;

exports[`MenuParent renders in open mode without children 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.span>
    File
  </styled.span>
  <styled.span>
    <styled.div>
      <styled.span>
        File
      </styled.span>
    </styled.div>
  </styled.span>
</styled.div>
`;
